reporter/sbom/spdx.go

```diff
@@ -0,0 +1,587 @@
+	packagePurposeApplication = "APPLICATION"
+	packagePurposeLibrary     = "LIBRARY"
+
+	packageAnnotatorTool = "Tool"
+	annotationOther      = "Other"
+
+	noneField = "NONE"
+
+	relationShipContains = common.TypeRelationshipContains
+	relationShipDepensOn = common.TypeRelationshipDependsOn
+
+	categoryPackageManager = common.CategoryPackageManager
+	packageManagerPURL     = common.TypePackageManagerPURL
+
+	categorySecurity  = common.CategorySecurity
+	securityCPE23Type = common.TypeSecurityCPE23Type
+	securityAdvisory  = common.TypeSecurityAdvisory
+)
+
+// ToSPDX converts ScanResult to SPDX Document
+func ToSPDX(r models.ScanResult) spdx.Document {
+	root := osToSpdxPackage(r)
+
+	packageToURLMap := createPackageToURLMap(r)
+	packages, relationships := spdxPackages(r, root, packageToURLMap)
+
+	doc := spdx.Document{
+		SPDXVersion:       spdx.Version,
+		DataLicense:       spdx.DataLicense,
+		SPDXIdentifier:    documentSPDXIdentifier,
+		DocumentName:      root.PackageName,
+		DocumentNamespace: fmt.Sprintf("%s-%s-%s", root.PackageName, root.PackageVersion, uuid.NewString()),
+		CreationInfo:      spdxCreationInfo(r),
+		Packages:          packages,
+		Relationships:     relationships,
+	}
+
+	sortSDPXDocument(&doc)
+
+	return doc
+}
+
+// SerializeSPDX serializes SPDX Document to JSON
+// Only supports JSON format
+func SerializeSPDX(doc spdx.Document) ([]byte, error) {
+	return json.MarshalIndent(&doc, "", "  ")
+}
+
+func osToSpdxPackage(r models.ScanResult) *spdx.Package {
+	family := constant.ServerTypePseudo
+	if r.Family != "" {
+		family = r.Family
+	}
+
+	var annotations []spdx.Annotation
+	annotations = appendAnnotation(annotations, "OsFamily", family, r.ReportedAt)
+	if r.RunningKernel.Release != "" {
+		annotations = appendAnnotation(annotations, "RunningKernelRelease", r.RunningKernel.Release, r.ReportedAt)
+	}
+	if r.RunningKernel.Version != "" {
+		annotations = appendAnnotation(annotations, "RunningKernelVersion", r.RunningKernel.Version, r.ReportedAt)
+	}
+
+	return &spdx.Package{
+		PackageSPDXIdentifier:     calculateSDPXIDentifier(elementOperatingSystem),
+		PackageName:               family,
+		PackageVersion:            r.Release,
+		PackageDownloadLocation:   noneField,
+		Annotations:               annotations,
+		PackageExternalReferences: nil,
+		PrimaryPackagePurpose:     packagePurposeOS,
+	}
+}
+
+func spdxCreationInfo(result models.ScanResult) *spdx.CreationInfo {
+	toolName := creatorTool
+	if result.ReportedVersion != "" {
+		toolName = fmt.Sprintf("%s-%s", creatorTool, result.ReportedVersion)
+	}
+
+	ci := spdx.CreationInfo{
+		Creators: []common.Creator{
+			{Creator: creatorOrganization, CreatorType: "Organization"},
+			{Creator: toolName, CreatorType: "Tool"},
+		},
+		Created: result.ReportedAt.Format(time.RFC3339),
+	}
+
+	return &ci
+}
+
+func spdxPackages(result models.ScanResult, root *spdx.Package, packageToURLMap map[string][]string) ([]*spdx.Package, []*spdx.Relationship) {
```

Comment:
I'm not sure whether an element can be nil or not... I prefer []spdx.Package if all elements are not nil.  How do you think?